### PR TITLE
[main-md] Rust for Substrate WIP

### PR DESCRIPTION
**Pull request by lisa-parity**

### CODE REVIEW

File: content/md/en/docs/main-docs/fundamentals/rust-basics.md [hunk 1]
```diff
@@ -1,230 +1,125 @@
...
 --- 
 
-_The goal of this article is to go over some of Rust's programming paradigms and how they're used in Substrate._
-_In doing so, it explains how Substrate uses no_std and why and the role of generics in Substrate runtimes_.
+[Rust](https://www.rust-lang.org/) is a highly performant, type sound, and memory efficient programming language.
+When you write programs in Rust, the Rust compiler checks your code before producing an executable, reducing the likelihood that your code will introduce memory leaks or other errors.
 
-The [Rust programming language](https://www.rust-lang.org/) enables writing software with a high degree of performance, type safety and memory efficiency.
-It is a compiled language, meaning that the compiler checks the code before an executable is produced, helping developers gain confidence in the code they write and making it difficult to write code with memory or concurency bugs.
+Substrate takes advantage of some key features of the Rust programming language to provide a flexible framework for creating mission-critical software with fewer errors.
```
Comment (sacha-l) [inline content/md/en/docs/main-docs/fundamentals/rust-basics.md:10]:
```suggestion
Substrate takes advantage of some key features of the Rust programming language to provide a flexible framework for creating mission-critical software.
```

File: content/md/en/docs/main-docs/fundamentals/rust-basics.md [hunk 2]
```diff
@@ -1,230 +1,125 @@
...
 
-## Programming paradigms
+One of the powerful features of Rust that Substrate takes full advantage of is the Rust **trait** system.
+In Rust, traits define what can be done by different types and how different types can interact with each other.
+Traits enable you to define shared behavior in an abstract way.
+You can then use **trait bounds** to specify that a **generic type** can be any type that has certain behavior.
 
-Rust has a sophisticated [trait system](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html) that helps developers make use of Substrate’s many layers of abstractions.
-The core features available to build these layers of abstraction are owed to Rust's system of traits and generics.
+The trait system enables different types to call the same set of methods to return different results.
```
Comment (sacha-l) [inline content/md/en/docs/main-docs/fundamentals/rust-basics.md:21]:
```suggestion
Rust's trait system enables different types to call the same set of methods to return different results.
```

File: content/md/en/docs/main-docs/fundamentals/rust-basics.md [hunk 3]
```diff
@@ -1,230 +1,125 @@
...
+Substrate uses the trait system to specify the set of operations that can be performed on a generic type.
+Because the behaviors are defined on a generic type, you can reuse and adapt domain-specific logic by modifying the type bounds to suit a specific context.
```
Comment (sacha-l) [inline content/md/en/docs/main-docs/fundamentals/rust-basics.md:23]: This isn't unique to Substrate, just a Rust thing.

File: content/md/en/docs/main-docs/fundamentals/rust-basics.md [hunk 4]
```diff
@@ -1,230 +1,125 @@
...
 
-### Generic types 
+## Function calls and arguments
 
-Rust's rich type system allows you to explicitly express your intent.
-For instance, the `Result` type ensures that every function call has an error case &mdash; a core principle for writing runtimes with Substrate.
-Similarly, function arguments can take the `Option` type, where the argument can be some value or `None` and both cases must be handled.
+In Rust, you can use the type system to enforce good coding practices.
+For example, you can use the `Result` type to ensure that every function call includes an error case.
+If you don't provide error handling, the Rust compiler issues a warning to remind you to add it.
```
Comment (sacha-l) [inline content/md/en/docs/main-docs/fundamentals/rust-basics.md:72]: It just won't compile.

File: content/md/en/docs/main-docs/fundamentals/rust-basics.md [hunk 5]
```diff
@@ -1,230 +1,125 @@
...
-However, the "unknown" parts of this Wasm target enforces the notion of making zero assumptions about the target environment, which is a key design decision in Substrate.
-
-It is also worth mentioning that there is `std` support for Substrate's Wasm target. 
-But this is not something that Wasm runtimes in Substrate support as it could open up unwanted errors such as leaking `std` code into Wasm environments, which will cause the system to panic.
-In addition, the `wasm32-unknown-unknown` target architecture and `no_std` have the same fundamental assumptions, making `no_std` a natural fit.
+## Macros and metaprogramming
 
-Rust `std` features are generally not compatible with the intended constraints of a Wasm runtime.
-For example, developers who attempt operations that are not allowed in the runtime, such as printing some text using `std`, could at worst cause the runtime to panic and terminate immediately.
+As in most programming languages, Rust macros encapsulate code that writes other code.
```
Comment (sacha-l) [inline content/md/en/docs/main-docs/fundamentals/rust-basics.md:103]:
```suggestion
Put simply, Rust macros are single lines of code that writes other code.
```

**Review by sacha-l — COMMENTED** [review]
Minor revision. I think we can rely even more on Rust resources and provide entry points to substrate without getting lost in the weeds of traits and generics.

File: content/md/en/docs/main-docs/fundamentals/rust-basics.md [hunk 6]
```diff
@@ -1,230 +1,129 @@
...
 featured_image:
 --- 
 
-_The goal of this article is to go over some of Rust's programming paradigms and how they're used in Substrate._
-_In doing so, it explains how Substrate uses no_std and why and the role of generics in Substrate runtimes_.
+[Rust](https://www.rust-lang.org/) is a highly performant, memory safe and efficient programming language and a first choice for Substrate for the following reasons:
 
-The [Rust programming language](https://www.rust-lang.org/) enables writing software with a high degree of performance, type safety and memory efficiency.
-It is a compiled language, meaning that the compiler checks the code before an executable is produced, helping developers gain confidence in the code they write and making it difficult to write code with memory or concurency bugs.
+- Rust is friendly: it's compiled, meaning that the compiler won’t ever let you write unsafe code that could crash your program at runtime.
```
Comment (sacha-l) [inline content/md/en/docs/main-docs/fundamentals/rust-basics.md:9]:
```suggestion
- Rust has a friendly compiler: it checks each and every variable you use and every memory address you reference to make it hard for you to write poor quality code.
```

File: content/md/en/docs/main-docs/fundamentals/rust-basics.md [hunk 7]
```diff
@@ -107,132 +107,19 @@ In Rust, there are declarative macros and three kinds of procedural macros:
...
-## Learn more
-
-- [Rust book](https://doc.rust-lang.org/book/)
-- [How rustup works](https://rust-lang.github.io/rustup/concepts/index.html)
-- [Why Rust?](https://www.parity.io/blog/why-rust)
-- [Cargo and crates.io](https://doc.rust-lang.org/book/ch14-00-more-about-cargo.html)
-- [Rust blog post](https://thenewstack.io/rust-by-the-numbers-the-rust-programming-language-in-2021/)
-- [Why Rust for smart contracts?](https://paritytech.github.io/ink-docs/why-rust-for-smart-contracts)
+* [Rust book](https://doc.rust-lang.org/book/)
+* [Why Rust?](https://www.parity.io/blog/why-rust)
```
Comment (NukeManDan) [inline content/md/en/docs/main-docs/fundamentals/rust-basics.md:74]:
```suggestion
* [Why Rust [for Parity and Substrate]?(https://www.parity.io/blog/why-rust)
```

**Pull request merged**